Codomain maps render url

will-moore opened the pull request:
# What this PR does

Updates the url query string for encoding codomain 'maps'.
See https://trello.com/c/PzMTIpY2/266-rendering-settings-in-url and original PR: https://github.com/openmicroscopy/openmicroscopy/pull/4839

This makes the encoding of 'maps' more extensible for addition of more maps in the future.
We now encode maps as a list of dictionaries. Each dictionary applies to a channel and each dictionary has a key that is an identifier for the map, and the value is another dictionary.
```"reverse"``` codomain only has an "enabled" boolean. For a 2-channel image this looks like:
```
?maps=[{"reverse":{"enabled":true}},{"reverse":{"enabled":false}}
```

# Testing this PR

1. Test rendering of images in Preview and Full viewer. Turn on/off reverse intensity and check that this is rendered correctly.

2. Copy & paste settings with Reverse-Intensity in various combinations of 
     - Copy via 'Copy' button in Preview panel OR via right-click menu on jsTree OR in full viewer
     - Paste via 'Paste' button in Preview panel OR via 'Paste and Save' on jsTree OR in full viewer

# Related reading

Link to cards, tickets, other PRs:

1. https://trello.com/c/PzMTIpY2/266-rendering-settings-in-url

2. Discussion at https://github.com/openmicroscopy/openmicroscopy/pull/4839

cc @chris-allan @waxenegger

will-moore committed:
Use ?maps=[{'reverse':{'enabled':true}] to encode codomain maps

will-moore committed:
viewport.getQuery() generates 'map=[]' query for reverseIntensity

will-moore committed:
Handle copy/paste rendering setting 'maps'

will-moore committed:
Handle 'maps' when 'paste' rdef after 'copy' fromid

will-moore commented:
@jburel I'll fix Travis, add tests and re-list... Doesn't need reviewing today (unless feedback on description)

will-moore committed:
Add robot test 'Test LUTs Reverse Intensity'

will-moore committed:
Fix 'Paste' rdef 'fromid' -> Preview panel

will-moore committed:
Merge remote-tracking branch 'origin/develop' into codomain_maps_render_url

will-moore committed:
flake8 fixes

will-moore committed:
Fix merge commit

waxenegger commented:
Tested on http://web-dev-merge.openmicroscopy.org.

In general it worked well, both when applying and copy and paste and in preview as well as full viewer (incl multiple selection and tree).

I have 2 cases, however, when a js error is caused and I suspect it's for the same reason. In both cases the rendering settings are supplied via the url that has all the parameters including map:

1) Hit the button "Full Viewer" in the preview tab then that image is opened with the "full" url.
2) Get the link to the image and open it in a new tab

Seems that the contents of the map param arrive at the following line as an unparsed string but it wants to work with the parsed array content already:

![image](https://cloud.githubusercontent.com/assets/1559229/22636713/eb518eb8-ec88-11e6-9ad7-cff2785ab9be.png)

will-moore committed:
Fix robot failing 'Test LUTs Reverse Intensity'

will-moore committed:
Fix opening of viewer from Preview panel with ?maps=

will-moore committed:
New robot test Test Launch Full Viewer

will-moore commented:
@waxenegger Thanks. Fixed and added a robot test for that.

waxenegger commented:
Yes, that fixed it, opening with a url and the "Full viewer" button works fine now.

![codomain_url](https://cloud.githubusercontent.com/assets/1559229/22681627/f8ec71ba-ed59-11e6-81d1-488039298491.png)

Though the robot doesn't seem to like your test case: https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/513/robot/

will-moore committed:
Fix failing robot test: Reverse intensity 'true' for channel 0

will-moore commented:
Last commit should fix [currently failing Test Launch Full Viewer](https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/516/robot/)

mtbc commented:
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/517/robot/ shows "Test Launch Full Viewer" passing and "Test Rating" newly failing.

will-moore committed:
Alert 'Setting image defaults failed' with more details.

To help with fixing error in robot tests with Firefox

will-moore committed:
Improve saveImgRDef error message

This allows us to know whether success is false or if response is false

will-moore committed:
Remove 'reset rdef' step at end of robot 'Test Rdef Save'

This was causing failures in Firefox when the next text started because the 'Save rdef' ajax
call was still in progress, hence returned with 'error' and caused alert.
See PR #5082.
We don't need to reset at the end of test since we now do it at the start which is
more robust (if previous test failed midway through

will-moore commented:
Last commit should fix failing robot test of Firefox at https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/520/robot/Web%20&%20Web/Web_1/Rdef%20Test/Test%20LUTs%20Reverse%20Intensity/

jburel commented:
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/521/robot/Web%20&%20Web/Web_1/Rdef%20Test/Test%20LUTs%20Reverse%20Intensity/ is now passing

jburel commented on components/tests/ui/testcases/web/rdef_test.txt:
```diff
@@ -373,7 +409,101 @@ Test Rdef Save
+    ${imageId2}=                            Get Selected Id From Tree
+    ${imgNodeId2}=                          Wait For Image Node         ${imageId2}
+    Right Click Rendering Settings          ${imgNodeId2}               Paste and Save
+    Wait For Image Src                      16_colors
```
did you check with invalid lut name?
Do we correctly fall back on color?

jburel commented:
Have you try passing an invalid values in the url e.g. lut=foo?

will-moore committed:
Add 'maps=' to end of query string instead of beginning

jburel commented:
The change will be "less" disruptive
Tests are passing

waxenegger commented:
yes, tests are passing (https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/522/console) and the maps parameter is at the end of the requests on http://web-dev-merge.openmicroscopy.org

waxenegger commented:
I might not be a bad idea to wrap the JSON.parse call in a try catch in case users go "evil" or map contents are accidentally messed up programmatically, just so that the image request remains valid. Of course one could always try things to break the url in other places if one wants to...

![image](https://cloud.githubusercontent.com/assets/1559229/23120641/ebae064a-f7a8-11e6-9f83-6687a2c9e2eb.png)

will-moore commented:
@waxenegger Yes, I guess JSON.parse is more sensitive to errors than the rest of the render string so I'll add a try/catch.

will-moore committed:
Use try/catch for JSON.parse of ?maps query

will-moore commented:
@waxenegger That last commit should give you an alert with invalid json for ```?maps=```

waxenegger commented:
I'm thinking 'alert' is perhaps a bit on the nose, I find the console log functions less disruptive to the user while still being helpful to the developers. Of course one could say the user might want to know that there is something wrong with the url so it boils down to premise/preference. 

Other than that everything works fine and the corresponding robot test is passing as well.

jburel commented:
We can downgrade if people complain and think ``alert`` is too much

Ideally we should reset the url to its "working" state but not a blocker
